src/com/xilinx/rapidwright/util/PerformanceExplorer.java

```diff
@@ -288,6 +301,38 @@ public void setEnsureExternalRoutability(boolean ensureExternalRoutability) {
         this.ensureExternalRoutability = ensureExternalRoutability;
     }
 
+    public String getExternalRoutabilitySideFile() {
+        return externalRoutabilitySideFile;
+    }
+
+    public void setExternalRoutabilitySideFile(String externalRoutabilitySideFile) {
+        this.externalRoutabilitySideFile = externalRoutabilitySideFile;
+    }
+
+    public Map<EDIFPort, PBlockSide> getExternalRoutabilitySideMap(EDIFNetlist netlist) {
+        if (externalRoutabilitySideMap == null) {
+            externalRoutabilitySideMap = new HashMap<>();
+            List<String> lines = FileTools.getLinesFromTextFile(getExternalRoutabilitySideFile());
+
+            for (String line : lines) {
+                String[] splitLine = line.split("\\s+");
+                String portRegex = splitLine[0];
+                String pblockSide = splitLine[1].toUpperCase();
+                for (EDIFPort port : netlist.getTopCell().getPorts()) {
+                    if (port.getBusName().matches(portRegex) ||
+                            port.getName().matches("\\" + EDIFTools.VIVADO_PRESERVE_PORT_INTERFACE + portRegex)) {
```

Comment:
Just an FYI - `String.matches()` is notoriously slow.  I think in this case the regex is unique per line, so refactoring it to compile the regex ahead of time is not going to help.  The performance of this method is probably not going to be an issue, so I don't think a change is needed here.